ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,467 @@
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    marker::PhantomData,
+    sync::{Arc, Mutex, Weak},
+};
+
+use async_trait::async_trait;
+use bitvec::{array::BitArray, prelude::Lsb0, slice::BitSlice};
+
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        context::{
+            dzkp_malicious::DZKPUpgraded as MaliciousDZKPUpgraded,
+            dzkp_semi_honest::DZKPUpgraded as SemiHonestDZKPUpgraded, Base, Context,
+            DZKPUpgradableContext, MaliciousContext, SemiHonestContext,
+        },
+        step::Gate,
+        RecordId,
+    },
+};
+
+/// `UnverifiedValues` are intermediate values that occur during a multiplication.
+/// These values need to be verified since there might have been malicious behavior.
+/// For a multiplication, let `(x1, x2), (x2, x3), (x3,x1)`, `(y1, y2), (y2, y3), (y3,y1)` be the shares of `x` and `y`
+/// and let the goal of the multiplication be to compute `x*y`.
+/// `(pi, p(i+1)` is the randomness from `PRSS` used for the multiplication by helper `i`.
+/// `z(i+1)` is the result of the multiplication received from the helper on the right.
+/// We do not need to store `zi` since it can be computed from the other stored values.
```

Comment:
This is a tradeoff. Just storing it would be simpler, but require more memory.

Reply:
yes, I think either way, it is simple since z can be easily computed from the other values when needed. I have not implemented the conversion yet. I think storing nor recomputing it adds significant overhead.